src/hotspot/share/services/virtualMemoryTracker.cpp

```diff
@@ -668,8 +668,11 @@ bool VirtualMemoryTracker::walk_virtual_memory(VirtualMemoryWalker* walker) {
       }
       head = head->next();
     }
-   }
-  return true;
+    return true;
+  } else {
+    assert(MemTracker::tracking_level() < NMT_summary, "Must be");
```

Comment:
Is this safe? We don't guard downgrading the tracking level with ThreadCritical.

Reply:
I think it is safe. There is a full fence before downgrading transition, therefore, it has a inter-thread happen-before relationship. So, if a thread sees transition (e.g. reset internal structure) with `ThreadCritical` lock held, it should see the downgraded tracking level.